### PR TITLE
Allow apply-cpp-change-to for other file extensions as well

### LINKED ISSUE #304: apply-cpp-change-to does not work for C source files
I tried to use the modifier `apply-cpp-change-to`, this is its documentation:
> apply-cpp-change-to: Add a function to a C/C++ source file. Each iteration adds a new function and removes the function added by the previous iteration. 

But when I specify a scenario with for example `apply-cpp-change-to = ["app/src/cpp/library.c"]` and try to run it, I get this error: 

```
* Settings
Project dir: ..
Output dir: profile-out-16
Profiler: none
Benchmark: true
Versions: []
Gradle User Home: gradle-user-home
Targets: [incremental_abi_change]
java.lang.RuntimeException: Could not create instance of mutator ApplyChangeToNativeSourceFileMutator
        at org.gradle.profiler.mutations.FileChangeMutatorConfigurator.getBuildMutatorForFile(FileChangeMutatorConfigurator.java:42)
        at org.gradle.profiler.mutations.FileChangeMutatorConfigurator.configure(FileChangeMutatorConfigurator.java:27)
        at org.gradle.profiler.ScenarioLoader.lambda$loadScenarios$2(ScenarioLoader.java:216)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
        at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
        at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
        at org.gradle.profiler.ScenarioLoader.loadScenarios(ScenarioLoader.java:218)
        at org.gradle.profiler.ScenarioLoader.doLoadScenarios(ScenarioLoader.java:156)
        at org.gradle.profiler.ScenarioLoader.loadScenarios(ScenarioLoader.java:137)
        at org.gradle.profiler.Main.run(Main.java:49)
        at org.gradle.profiler.Main.main(Main.java:18)
Caused by: java.lang.IllegalArgumentException: Can only modify C++ source or header files
        at org.gradle.profiler.mutations.ApplyChangeToNativeSourceFileMutator.<init>(ApplyChangeToNativeSourceFileMutator.java:12)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:64)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:481)
        at org.gradle.profiler.mutations.FileChangeMutatorConfigurator.getBuildMutatorForFile(FileChangeMutatorConfigurator.java:37)
        ... 16 more
```

library.c is a file in our project that's used with JNI. It looks about like this:
```c
#include <jni.h>

JNIEXPORT jstring JNICALL
Java_de_maibornwolff_sample_data_xyz_XyzLibrary_getXyz(JNIEnv *env, jobject thiz) {
    return (*env)->  NewStringUTF(env, "sample data");
}

//... more JNIEXPORT-marked functions
```

I then renamed the file to `.cpp` and then `gradle-profiler` did actually mutate the file. It added `int _m_386a15b3_dc61_408a_8c5e_054204298a4e_incremental_abi_change_7a762916_WARM_UP_1 () { }` which is expected.

### PR DESCRIPTION
**Pull request by carstenhag**
Fixes #304 
Use list of valid c/c++ source and header files instead of few hardcoded values 🥇 
Also added further tests.

### CODE REVIEW

**Comment by carstenhag** [comment]
Just noticed that the tests don't run, it's missing from the pipeline it seems? :O

**Comment by carstenhag** [comment]
I saw that two builds failed, but unfortunately I don't know what is causing this. @lptr can you please chime in? Likely it's unrelated, seems to be something teamcity-specific?

**Comment by lptr** [comment]
Yes, it looks like our TeamCity has some issues. We'll need to investigate further, sorry.

**Comment by wolfs** [comment]
@carstenhag It looks like your branch is based on an older version of gradle profiler (November 2019). Can you rebase it on current master? That should fix the CI problems.

**Comment by carstenhag** [comment]
@wolfs Yep that was it, I had forked/cloned the repo a while ago and didnt have the latest upstream commits. Thanks for clearing it up!

File: src/main/java/org/gradle/profiler/mutations/ApplyChangeToNativeSourceFileMutator.java [hunk 1]
```diff
@@ -1,22 +1,42 @@
...
+
+    private final List<String> nativeHeaderFileEndings = Arrays.asList(
+        ".h", ".hh", ".H", ".hp", ".hxx", ".hpp", ".HPP", ".h++", ".tcc"
+    );
+
     public ApplyChangeToNativeSourceFileMutator(File file) {
         super(file);
-        if (!sourceFile.getName().endsWith(".cpp") && !sourceFile.getName().endsWith(".h") && !sourceFile.getName().endsWith(".hpp")) {
-            throw new IllegalArgumentException("Can only modify C++ source or header files");
+        String fileExtension = "." + FilenameUtils.getExtension(sourceFile.getName());
```
Comment (wolfs) [inline src/main/java/org/gradle/profiler/mutations/ApplyChangeToNativeSourceFileMutator.java:25]: I would drop the `.` from the extensions, so you don't have to add it here.

File: src/main/java/org/gradle/profiler/mutations/ApplyChangeToNativeSourceFileMutator.java [hunk 2]
```diff
@@ -1,22 +1,42 @@
...
+        List<String> extensionList = Stream.concat(nativeSourcecodeFileEndings.stream(), nativeHeaderFileEndings.stream())
```
Comment (wolfs) [inline src/main/java/org/gradle/profiler/mutations/ApplyChangeToNativeSourceFileMutator.java:26]:
```suggestion
        boolean isSupportedExtension = Stream.concat(nativeSourcecodeFileEndings.stream(), nativeHeaderFileEndings.stream()).anyMatch(fileExtension::equals)
```

How about ☝️ , then you don't need to create the extra list.

File: src/main/java/org/gradle/profiler/mutations/ApplyChangeToNativeSourceFileMutator.java [hunk 3]
```diff
@@ -1,22 +1,42 @@
...
 import java.io.File;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 public class ApplyChangeToNativeSourceFileMutator extends AbstractFileChangeMutator {
 
+    // Both lists taken from https://gcc.gnu.org/onlinedocs/gcc/Overall-Options.html
+    private final List<String> nativeSourcecodeFileEndings = Arrays.asList(
```
Comment (wolfs) [inline src/main/java/org/gradle/profiler/mutations/ApplyChangeToNativeSourceFileMutator.java:15]:
```suggestion
    private static final List<String> nativeSourcecodeFileEndings = ImmutableSet.of(
```

File: src/main/java/org/gradle/profiler/mutations/ApplyChangeToNativeSourceFileMutator.java [hunk 4]
```diff
@@ -1,22 +1,42 @@
...
+        ".c", ".cc", ".cp", ".cxx", ".cpp", ".CPP", ".c++", ".C"
+    );
+
+    private final List<String> nativeHeaderFileEndings = Arrays.asList(
```
Comment (wolfs) [inline src/main/java/org/gradle/profiler/mutations/ApplyChangeToNativeSourceFileMutator.java:19]:
```suggestion
    private static final List<String> nativeHeaderFileEndings = ImmutableSet.of(
```

**Review by wolfs — COMMENTED** [review]
Thank you for your contribution. I think the basic idea for the change is good! I left some comments regarding the implementation.

File: src/main/java/org/gradle/profiler/mutations/ApplyChangeToNativeSourceFileMutator.java (hunk 1 above)
Comment (carstenhag) [reply src/main/java/org/gradle/profiler/mutations/ApplyChangeToNativeSourceFileMutator.java]: Agree, thanks :)

File: src/main/java/org/gradle/profiler/mutations/ApplyChangeToNativeSourceFileMutator.java (hunk 2 above)
Comment (carstenhag) [reply src/main/java/org/gradle/profiler/mutations/ApplyChangeToNativeSourceFileMutator.java]: Ahh yeah way better, I usually develop in Kotlin so I wasn't aware of the java way of doing this.

File: src/main/java/org/gradle/profiler/mutations/ApplyChangeToNativeSourceFileMutator.java [hunk 5]
```diff
@@ -1,22 +1,42 @@
...
-        if (!sourceFile.getName().endsWith(".cpp") && !sourceFile.getName().endsWith(".h") && !sourceFile.getName().endsWith(".hpp")) {
-            throw new IllegalArgumentException("Can only modify C++ source or header files");
+        String fileExtension = FilenameUtils.getExtension(sourceFile.getName());
+        boolean isSupportedExtension = Stream.concat(nativeSourcecodeFileEndings.stream(), nativeHeaderFileEndings.stream())
+            .anyMatch(fileExtension::equals);
+
+        if (isSupportedExtension) {
+            return;
         }
+
```
Comment (carstenhag) [inline src/main/java/org/gradle/profiler/mutations/ApplyChangeToNativeSourceFileMutator.java:31]:
```suggestion
```

**Comment by wolfs** [comment]
@carstenhag: FYI: I simplified it a little further after merging: https://github.com/gradle/gradle-profiler/commit/3bdaef7151013bddf4a42845b0381293aa613093

**Pull request merged**